### PR TITLE
Alternative

### PR DESCRIPTION
**Pull request by stephenjudkins**

### CODE REVIEW

**Comment by stephenjudkins** [comment]
As @mpilquist confirmed, `some` and `many` are undecidable for  `Alternative` instances for `Option` and `List`, and I believe they're also undecidable for all plausible instances that might be in `cats.std`.

Given that, `Alternative` still provides some useful laws even if it doesn't include any derived functions. Anyone have a problem with merging what's here now?

File: laws/src/main/scala/cats/laws/AlternativeLaws.scala [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import cats.Alternative
+
+import cats.syntax.all._
+
+trait AlternativeLaws[F[_]] extends ApplicativeLaws[F] with MonoidKLaws[F] {
+  implicit override def F: Alternative[F]
+  implicit def algebra[A] = F.algebra[A]
+
+  def rightAbsorption[A, B](ff: F[A => B]): IsEq[F[B]] =
+    (F.empty[A] apply ff) <-> F.empty[B]
```
Comment (fthomas) [inline laws/src/main/scala/cats/laws/AlternativeLaws.scala]: Minor nitpick: could you prepend all laws with `alternative`? This way it is easier to see where these laws are coming from if `AlternativeLaws` is mixed into another law trait.

File: laws/src/main/scala/cats/laws/discipline/AlternativeTests.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import cats.laws.AlternativeLaws
+import org.scalacheck.Arbitrary
+import org.scalacheck.Prop._
+
+trait AlternativeTests[F[_]] extends ApplicativeTests[F] with MonoidKTests[F]  {
+  def laws:AlternativeLaws[F]
+
+  def alternative[A: Arbitrary, B: Arbitrary, C: Arbitrary](implicit
+    ArbF: ArbitraryK[F],
+    arfFAB: Arbitrary[F[A => B]],
```
Comment (fthomas) [inline laws/src/main/scala/cats/laws/discipline/AlternativeTests.scala]: Could you rename that to `ArbFAB`?

**Comment by fthomas** [comment]
:+1: looks good to me now, thanks

**Comment by xuwei-k** [comment]
why not `MonadCombine[F[_]] extends Alternative[F]` ?

**Comment by non** [comment]
I agree with @xuwei-k, we should update `MonadCombine` now that we have `Alternative`.

**Comment by rossabaker** [comment]
And also `MonadCombineLaws` and `MonadCombineTests`.

**Comment by ceedubs** [comment]
:+1: looks good to me

**Comment by rossabaker** [comment]
`MonadCombine` now also extends `MonoidK` via `Alternative`.  Is the direct extension intentional or cruft?

**Comment by ceedubs** [comment]
Oh good catch @rossabaker. I would think the `extends MonoidK` should be removed from `MonadCombine`

**Comment by rossabaker** [comment]
:+1:

File: laws/src/main/scala/cats/laws/AlternativeLaws.scala [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  implicit def algebra[A] = F.algebra[A]
+
+  def alternativeRightAbsorption[A, B](ff: F[A => B]): IsEq[F[B]] =
+    (F.empty[A] apply ff) <-> F.empty[B]
+
+  def alternativeLeftDistributivity[A, B](fa: F[A], fa2: F[A], f: A => B): IsEq[F[B]] =
+    ((fa |+| fa2) map f) <-> ((fa map f) |+| (fa2 map f))
+
+  def alternativeRightDistributivity[A, B](fa: F[A], ff: F[A => B], fg: F[A => B]): IsEq[F[B]] =
+    (fa apply (ff |+| fg)) <-> ((fa apply ff) |+| (fa apply fg))
```
Comment (xuwei-k) [inline laws/src/main/scala/cats/laws/AlternativeLaws.scala]: What is `right distributivity`? where came from?

it does not exists in Haskell https://hackage.haskell.org/package/base-4.8.0.0/docs/src/GHC-Base.html#Alternative

```
-- | A monoid on applicative functors.
--
-- If defined, 'some' and 'many' should be the least solutions
-- of the equations:
--
-- * @some v = (:) '<$>' v '<*>' many v@
--
-- * @many v = some v '<|>' 'pure' []@
```

and `XorT` does not satisfy this law.

- https://github.com/non/cats/pull/287
- https://github.com/non/cats/commit/c11a1bfe880b4e445e85cd2c34579

I think `right distributivity` is not necessary.

**Pull request merged**